src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenForEachTests.cs

```diff
@@ -5327,5 +5327,125 @@ public static class Extensions
 }";
             CompileAndVerify(source, parseOptions: TestOptions.Regular9, expectedOutput: "123123");
         }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/73741")]
+        public void MutatingThroughRefFields_01()
+        {
+            var source = """
+                using System;
+
+                V[] arr = new V[3];
+
+                foreach (var r in new E(arr))
+                {
+                    r.V.F++;
+                }
+
+                foreach (var v in arr) Console.Write(v.F);
+
+                ref struct E(V[] arr)
+                {
+                    int i;
+                    public E GetEnumerator() => this;
+                    public R Current => new(ref arr[i - 1]);
+                    public bool MoveNext() => i++ < arr.Length;
+                }
+
+                ref struct R(ref V v)
+                {
+                    public ref V V = ref v;
```

Comment:
The submitted repro has these as `readonly ref` fields. Let's add some more tests with different variants of `readonly`ness sprinkled throughout. It would also be good to test variables that come from `ref foreach` loops.